torch/_meta_registrations.py

```diff
@@ -529,3 +525,56 @@ def meta_embedding_bag_forward_only(weight, indices, offsets, *args):
     if offsets.device.type == "cpu":
         bag_size = offsets.new_empty(offsets.size())
     return output, offset2bag, bag_size, max_indices
+
+
+def _get_reduction_dtype(input, dtype, promote_int_to_long=True):
+    # if specified, dtype takes precedence
+    if dtype:
+        return dtype
+
+    if input.dtype.is_floating_point or input.dtype.is_complex:
+        return input.dtype
+    elif promote_int_to_long:
+        return torch.long
+
+    return input.dtype
+
+
+def _get_reduction_output_size(input, dims, keepdim):
+    if input.size() == () or tuple(dims) == ():
+        return ()
+
+    out_size = list(input.size())
+    dims = utils.canonicalize_dims(input.dim(), dims)
+    # Sort in descending order as we have to pop
+    # elements out of `out_size` (if keepdim = False)
+    for dim in sorted(dims, reverse=True):
+        if keepdim:
+            out_size[dim] = 1
+        else:
+            out_size.pop(dim)
+
+    return out_size
+
+
+@torch.library.impl(meta_lib, "nansum")
+@torch.library.impl(meta_lib, "nansum.out")
+@out_wrapper
+def meta_nansum(input, dims=(), keepdim=False, *, dtype=None):
+    output_dtype = _get_reduction_dtype(input, dtype, promote_int_to_long=True)
+    output_size = _get_reduction_output_size(input, dims, keepdim)
+    return input.new_empty(output_size, dtype=output_dtype)
+
+
+@torch.library.impl(meta_lib, "nanmedian")
+def meta_nanmedian(input):
+    output_size = _get_reduction_output_size(input, (), False)
+    return input.new_empty(output_size)
+
+
+@torch.library.impl(meta_lib, "nanmedian.dim_values")
+@torch.library.impl(meta_lib, "nanmedian.dim")
+@out_wrapper_multi("values", "indices")
+def meta_nanmedian_dim(input, dim=-1, keepdim=False):
+    output_size = _get_reduction_output_size(input, (dim,), keepdim)
+    return input.new_empty(output_size), input.new_empty(output_size, dtype=torch.long)
```

Comment:
nit - `new_empty` sets the `requires_grad` to `False` by default but we should propagate it from input right?

Reply:
It does set it to `False` (see below). Interestingly, none of the registration pass `requires_grad` (and tests don't complain). Should we update all ops and update the test to verify if `requires_grad` is propagated?

```python
import torch
from torch._subclasses.fake_tensor import (
    FakeTensor,
    FakeTensorMode,
)

def _to_tensormeta(x):
    mode = torch._prims.utils.get_prim_fake_mode()
    if isinstance(x, torch.Tensor):
        out = FakeTensor.from_tensor(x, mode)
        return out
    return x

t = torch.randn((), requires_grad=True)
assert t.new_empty((2, 2)).requires_grad == t.requires_grad  # Fails 
```

Reply:
The meta functions live below autograd so it doesn't matter, you'll always be running below what is effectively `no_grad`